File: public/code/indicators__rsi_divergence.py
"""
RSI Divergence Strategy
=======================
Regular bullish: Fiyat LL, RSI HL = trend tersine dönecek (long).
Regular bearish: Fiyat HH, RSI LH = trend tersine dönecek (short).
Hidden bullish: Fiyat HL, RSI LL = trend devamı (long).
Hidden bearish: Fiyat LH, RSI HH = trend devamı (short).
"""
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
import pandas as pd
import numpy as np
from common import make_signal, rsi, atr, swing_highs_lows


def detect_rsi_divergence_signal(
    df: pd.DataFrame,
    rsi_period: int = 14,
    swing_lb: int = 5,
    lookback: int = 30,
) -> dict:
    if len(df) < lookback + rsi_period + 10:
        return make_signal(reason="Insufficient data")

    rsi_v = rsi(df["close"], rsi_period)
    sh, sl = swing_highs_lows(df, swing_lb)
    current = df.iloc[-1]
    current_atr = atr(df, 14).iloc[-1]

    # Son 2 swing high & low
    recent_lows = df["low"][sl.notna()].tail(2)
    recent_highs = df["high"][sh.notna()].tail(2)
    if len(recent_lows) < 2 or len(recent_highs) < 2:
        return make_signal(reason="Not enough swings")

    # Karşılık gelen RSI değerleri
    low_indices = sl.dropna().tail(2).index
    high_indices = sh.dropna().tail(2).index
    rsi_at_lows = [rsi_v.loc[i] for i in low_indices]
    rsi_at_highs = [rsi_v.loc[i] for i in high_indices]

    # Regular bullish divergence: price LL, RSI HL
    if recent_lows.iloc[1] < recent_lows.iloc[0] and rsi_at_lows[1] > rsi_at_lows[0] and rsi_at_lows[1] < 40:
        entry = current["close"]
        stop_loss = recent_lows.iloc[1] - 0.3 * current_atr
        risk = entry - stop_loss
        if risk > 0:
            return make_signal(
                signal="long",
                entry=entry,
                stop_loss=stop_loss,
                take_profit=[entry + risk * 2, entry + risk * 3, entry + risk * 5],
                confidence=0.78,
                reason="Regular bullish RSI divergence",
                metadata={"rsi_now": float(rsi_v.iloc[-1])},
            )

    # Regular bearish: price HH, RSI LH
    if recent_highs.iloc[1] > recent_highs.iloc[0] and rsi_at_highs[1] < rsi_at_highs[0] and rsi_at_highs[1] > 60:
        entry = current["close"]
        stop_loss = recent_highs.iloc[1] + 0.3 * current_atr
        risk = stop_loss - entry
        if risk > 0:
            return make_signal(
                signal="short",
                entry=entry,
                stop_loss=stop_loss,
                take_profit=[entry - risk * 2, entry - risk * 3, entry - risk * 5],
                confidence=0.78,
                reason="Regular bearish RSI divergence",
                metadata={"rsi_now": float(rsi_v.iloc[-1])},
            )

    return make_signal(reason="No RSI divergence")

File: public/code/indicators__rsi_divergence.ts
/**
 * RSI Divergence Strategy - TypeScript
 */
import { Candle, Signal, makeSignal, rsi, atr, swingHighs, swingLows } from "../../common";

export function detectRsiDivergenceSignal(
  candles: Candle[],
  opts: { rsiPeriod?: number; swingLb?: number } = {}
): Signal {
  const { rsiPeriod = 14, swingLb = 5 } = opts;
  if (candles.length < rsiPeriod + 30) return makeSignal({ reason: "Insufficient data" });

  const closes = candles.map((c) => c.close);
  const rsiV = rsi(closes, rsiPeriod);
  const sh = swingHighs(candles, swingLb);
  const sl = swingLows(candles, swingLb);

  const lowIndices: number[] = [];
  const highIndices: number[] = [];
  sl.forEach((v, i) => v !== null && lowIndices.push(i));
  sh.forEach((v, i) => v !== null && highIndices.push(i));

  if (lowIndices.length < 2 || highIndices.length < 2)
    return makeSignal({ reason: "Not enough swings" });

  const lastLow1 = lowIndices[lowIndices.length - 2];
  const lastLow2 = lowIndices[lowIndices.length - 1];
  const lastHigh1 = highIndices[highIndices.length - 2];
  const lastHigh2 = highIndices[highIndices.length - 1];

  const current = candles[candles.length - 1];
  const atrV = atr(candles, 14);
  const ai = atrV.length - 1;

  // Regular bullish: lower low in price, higher low in RSI
  if (
    candles[lastLow2].low < candles[lastLow1].low &&
    rsiV[lastLow2] > rsiV[lastLow1] &&
    rsiV[lastLow2] < 40
  ) {
    const entry = current.close;
    const stopLoss = candles[lastLow2].low - 0.3 * atrV[ai];
    const risk = entry - stopLoss;
    if (risk > 0)
      return makeSignal({
        signal: "long", entry, stop_loss: stopLoss,
        take_profit: [entry + risk * 2, entry + risk * 3, entry + risk * 5],
        confidence: 0.78, reason: "Regular bullish RSI divergence",
        timestamp: current.timestamp, metadata: { rsi_now: rsiV[ai] },
      });
  }

  if (
    candles[lastHigh2].high > candles[lastHigh1].high &&
    rsiV[lastHigh2] < rsiV[lastHigh1] &&
    rsiV[lastHigh2] > 60
  ) {
    const entry = current.close;
    const stopLoss = candles[lastHigh2].high + 0.3 * atrV[ai];
    const risk = stopLoss - entry;
    if (risk > 0)
      return makeSignal({
        signal: "short", entry, stop_loss: stopLoss,
        take_profit: [entry - risk * 2, entry - risk * 3, entry - risk * 5],
        confidence: 0.78, reason: "Regular bearish RSI divergence",
        timestamp: current.timestamp, metadata: { rsi_now: rsiV[ai] },
      });
  }
  return makeSignal({ reason: "No RSI divergence" });
}
